### DISCUSSION TITLE
What is the difference between `bin/console debug:container` and `bin/console debug:autowiring`?

### DISCUSSION THREAD

**Discussion by nbro** [symfony/symfony#54928] [Q&A]
If I understand correctly, both show services. 

`bin/console debug:container` should show all available services and I suppose that `bin/console debug:autowiring` only shows the ones that we can _inject_ as parameters into our methods (controllers in particular). Is this correct? If that's correct, then why aren't all services autowiring, i.e. why are there services which we can't simply inject and thus, I suppose, we need to manually create with `new`?

**Comment by ghost** [comment]
`debug:container` returns information about services inside DI container.
`debug:autowiring` returns information about classes and interfaces that DI container will inject automatically as constructor parameters without the need of configuring the constructor in the config file.

**Reply by nbro** [reply]
This doesn't answer my question - why aren't all services injectable? 

I already knew the definitions that you can find with the `--help` flag.

Which constructors are you talking about? What config file? Don't you usually inject the services directly into the controller functions? I suppose you can inject them in any function, though, including constructors.

**Comment by mbabker** [comment]
The `debug:container` command has ways to provide information about everything in your DI container.  Services, parameters, env vars, etc.  It is there to help you get information about what is in the compiled container.

The `debug:autowiring` command gives you hints about what the container knows how to autowire, essentially figuring out how to configure what gets injected into various places with zero configuration.  This autowiring applies to things like the services injected into controllers or class constructors.

> If that's correct, then why aren't all services autowiring, i.e. why are there services which we can't simply inject and thus, I suppose, we need to manually create with `new`?

First, autowiring applies to configuring the services in your DI container.  It's not a runtime behavior, and Symfony can't overload the PHP `new` operator to autowire something at runtime.  The point of autowiring is to make it so you spend less time manually configuring things in your `config/services.yaml` file, or using other ways of explicit manual configuration for the container (i.e. the various attributes the DI container knows how to parse).

Second, not everything can be autowired for one reason or another, or there might be an ambiguity when using autowiring and you'd need an extra bit of information to resolve that ambiguity.  Using the example right in the `bin/console debug:autowiring --help` output, if you run `bin/console debug:autowiring log`, you can see all of the autowiring aliases for every logger configured in your application.  If you've added extra logging channels to your application, and a service needs to use the logger for a specific channel, the info in that command's output tells you exactly how to typehint that in your parameter list so the autowiring logic in the container can work out "this service needs this specific logger" instead of providing the default logger.

**Reply by nbro** [reply]
> not everything can be autowired for one reason or another

It would be great if you could provide 1 or more examples of something that cannot be autowired and explain why.

>  autowiring applies to configuring the services in your DI container. It's not a runtime behavior, and Symfony can't overload the PHP new operator to autowire something at runtime

I did not say that `new` is overloaded. But it seems that the primary advantage of autowiring services is that we don't need to explicitly create service objects with `new` and that the container automatically creates the objects for us.

**Reply by MatTheCat** [reply]
> we don't need to explicitly create service objects with `new` […] the container automatically creates the objects for us.

This is [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) at work. Autowiring is a way to simplify the configuration from which the container will be built. You can read https://symfony.com/doc/current/service_container/autowiring.html which should answer your questions.
